### PR TITLE
Amplify CLI & SDKs Integration Test

### PR DESCRIPTION
**Pull request by AaronZyLee**
This version is only for triggering tests in React App. When script runs, it will do the following things:

- Backend:

1. Run amplify configuration.
2. Create project folder for amplify.
3. Git clone the sample app.
4. Run amplify init and add auth/storage/api for each test.
5. Check out the user pool and sign up a user.
6. Generate a test environment file.

- Frontend:

1. Copy `aws-exports.js` to the sample app folder.
2. Build the app.
3. Start the app.
4. Run tests for each category.
5. Close the app.

For each category, a specific port number is assigned, in order that the tests can run in parallel.

### CODE REVIEW

File: packages/amplify-ui-tests/__tests__/js/api.test.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import { createNewProjectDir, deleteProjectDir, createTestMetaFile } from '../../src/utils';
+import { addAuthWithDefault } from '../../src/categories/auth';
+import { copyAWSExportsToProj, existsAWSExportsPath } from '../../src/utils/projectMeta';
+import { runCypressTest, gitCloneSampleApp, buildApp, startServer, closeServer, signUpNewUser } from '../../src/utils/command';
+
+describe('Javascript SDK', () => {
+  let projRoot: string;
+  let destRoot: string;
+  const API_PORT_NUMBER: string = '3003';
+  const JS_SAMPLE_APP_REPO: string = 'https://github.com/AaronZyLee/photo-albums.git';
```
Comment (UnleashedMind) [inline packages/amplify-ui-tests/__tests__/js/api.test.ts]: update the repo url to the official one

File: packages/amplify-ui-tests/__tests__/js/auth.test.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import { addAuthWithDefault } from '../../src/categories/auth';
+import { existsAWSExportsPath, copyAWSExportsToProj} from '../../src/utils/projectMeta';
+import { runCypressTest, startServer, closeServer, gitCloneSampleApp, buildApp, signUpNewUser } from '../../src/utils/command'
+
+
+describe('Javascript SDK:', () => {
+  let projRoot: string;
+  let destRoot: string;
+  const AUTH_PORT_NUMBER: string = '3001';
+  const JS_SAMPLE_APP_REPO: string = 'https://github.com/AaronZyLee/photo-albums.git';
```
Comment (UnleashedMind) [inline packages/amplify-ui-tests/__tests__/js/auth.test.ts]: update the repo url to the official one when it's ready

File: packages/amplify-ui-tests/__tests__/js/storage.test.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import { createNewProjectDir, deleteProjectDir, createTestMetaFile } from '../../src/utils';
+import { addAuthWithDefault } from '../../src/categories/auth';
+import { copyAWSExportsToProj, existsAWSExportsPath } from '../../src/utils/projectMeta';
+import { runCypressTest, gitCloneSampleApp, buildApp, startServer, closeServer, signUpNewUser } from '../../src/utils/command';
+
+describe('Javascript SDK:', () => {
+    let projRoot: string;
+    let destRoot: string;
+    const STORAGE_PORT_NUMBER: string = '3002';
+    const JS_SAMPLE_APP_REPO: string = 'https://github.com/AaronZyLee/photo-albums.git';
```
Comment (UnleashedMind) [inline packages/amplify-ui-tests/__tests__/js/storage.test.ts]: update the repo url to the official one when it's ready

File: .circleci/config.yml [hunk 4]
```diff
@@ -224,3 +346,25 @@ workflows:
...
+      #       - build
+      # - integration_test_ios:
+      #     appname: PhotoAlbum
+      #     filters:
+      #       branches:
+      #         only:
+      #           - integtest
+      #     requires:
+      #       - build
+      - integration_test_android:
```
[the remark this answers is not in the document]
Comment (AaronZyLee) [reply .circleci/config.yml]: This is just for temporary use. I will uncomment those code before the merge

File: packages/amplify-ui-tests/Readme.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+Before running the scripts, you need to do the installation by the following steps:
+* Clone this repo. Run `npm run setup-dev` under the repo root.
+* Then `cd packages/amplify-ui-tests` to make sure you are under this package.
+
+## Configuration
+To run the tests locally, you need to have your AWS credentials stored in a `.env` file of this package. These values are used to configure the tests projects.
+
+To be more specific, you need two variables in `.env`, which are `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.
+
+The `.env` file does not get commited as its in `.gitignore` file.
```
[the remark this answers is not in the document]
Comment (AaronZyLee) [reply packages/amplify-ui-tests/Readme.md]: When doing `amplify configure`, the question "specify the aws region" does not support string input but using arrows to choose. So the default region is the first choice of region, in which we just press `Enter` for this question.

[the remark this answers is not in the document]
Comment (AaronZyLee) [reply packages/amplify-ui-tests/Readme.md]: That's why region is not included in `.env`. The region is required for `aws_cli` configuration. However, when running in local environment, I assume that user already has `aws_cli` installed and configured.

File: packages/amplify-ui-tests/Readme.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+## How to run the AWS resources provision scripts
+
+* Run `npm run config [your-project-root] [sdk] [list: categories]`. `[sdk]` includes `js`, `ios` and `android`. `[list: categories]` includes `auth`, `storage` and `api`
+* Before you run adding api for your cloud, paste your own graphql schema file under `./schemas`, change the file name to `simple_model.graphql`
```
[the remark this answers is not in the document]
Comment (AaronZyLee) [reply packages/amplify-ui-tests/Readme.md]: Sure. I will add the example part.

File: packages/amplify-ui-tests/src/utils/index.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  //return process.env.CI ? false : true;
+  return process.env.CI ? true : false;
+}
+
+export function getEnvVars(): { } {
+  return { ...process.env };
+}
+
+export function getSampleRootPath() : string {
+  //return join(__dirname, '../../../../..', 'amplify-js-samples-staging')
```
[the remark this answers is not in the document]
Comment (AaronZyLee) [reply packages/amplify-ui-tests/src/utils/index.ts]: I remove the whole method. This is not used in tests.

**Pull request merged**